include/deal.II/lac/vector.templates.h

```diff
@@ -614,6 +563,487 @@ namespace internal
+    do_vectorized(const size_type i) const
+    {
+      VectorizedArray<Number> x;
+      x.load(X+i);
+      return std::abs(x);
+    }
+
+    const Number *X;
+  };
+
+  template <typename Number, typename RealType>
+  struct NormP
+  {
+    static const bool vectorizes = VectorizedArray<Number>::n_array_elements > 1;
+
+    RealType
+    operator() (const size_type i) const
+    {
+      return std::pow(numbers::NumberTraits<Number>::abs(X[i]), p);
+    }
+
+    VectorizedArray<Number>
+    do_vectorized(const size_type i) const
+    {
+      VectorizedArray<Number> x;
+      x.load(X+i);
+      return std::pow(std::abs(x),p);
+    }
+
+    const Number *X;
+    RealType p;
+  };
+
+  template <typename Number>
+  struct MeanValue
+  {
+    static const bool vectorizes = VectorizedArray<Number>::n_array_elements > 1;
+
+    Number
+    operator() (const size_type i) const
+    {
+      return X[i];
+    }
+
+    VectorizedArray<Number>
+    do_vectorized(const size_type i) const
+    {
+      VectorizedArray<Number> x;
+      x.load(X+i);
+      return x;
+    }
+
+    const Number *X;
+  };
+
+  template <typename Number>
+  struct AddAndDot
+  {
+    static const bool vectorizes = VectorizedArray<Number>::n_array_elements > 1;
+
+    Number
+    operator() (const size_type i) const
+    {
+      X[i] += a * V[i];
+      return X[i] * Number(numbers::NumberTraits<Number>::conjugate(W[i]));
+    }
+
+    VectorizedArray<Number>
+    do_vectorized(const size_type i) const
+    {
+      VectorizedArray<Number> x, w, v;
+      x.load(X+i);
+      v.load(V+i);
+      x += a * v;
+      x.store(X+i);
+      // may only load from W after storing in X because the pointers might
+      // point to the same memory
+      w.load(W+i);
+      return x * w;
+    }
+
+    Number *X;
+    const Number *V, *W;
+    Number a;
+  };
+
+
+
+  // this is the inner working routine for the accumulation loops
+  // below. This is the standard case where the loop bounds are known. We
+  // pulled this function out of the regular accumulate routine because we
+  // might do this thing vectorized (see specialized function below)
+  template <typename Operation, typename ResultType>
+  void
+  accumulate_regular(const Operation &op,
+                     size_type       &n_chunks,
+                     size_type       &index,
+                     ResultType (&outer_results)[128],
+                     internal::bool2type<false>,
+                     const unsigned int start_chunk=0)
+  {
+    AssertIndexRange(start_chunk, n_chunks+1);
+    for (size_type i=start_chunk; i<n_chunks; ++i)
+      {
+        ResultType r0 = op(index);
+        ResultType r1 = op(index+1);
+        ResultType r2 = op(index+2);
+        ResultType r3 = op(index+3);
+        index += 4;
+        for (size_type j=1; j<8; ++j, index += 4)
+          {
+            r0 += op(index);
+            r1 += op(index+1);
+            r2 += op(index+2);
+            r3 += op(index+3);
```

Comment:
This is actually interesting. I believe this code was written back in the 32 bit days of x86. Back then, there were few fp registers. I wonder if one could do this more efficient today with a loop that is unrolled 8 times, for example. How many double precision fp registers do we have today on x86-64? 

@kalt - any opinions?